packages/belos/tpetra/test/MINRES/test_minres_diag.cpp

```diff
@@ -0,0 +1,435 @@
+  n_global = n_in;
+#endif
+  pMap =  Teuchos::rcp( new MP(n_global, n_in, 0, pComm) );
+
+  pPE = Teuchos::rcp( new Trilinos_Interface<OP, ST, MP, MV>(pA, pComm, pMap ) );
+
+  pProb = Teuchos::rcp( new Belos::LinearProblem<double,MV,OP>() );
+  pProb->setOperator( pPE );
+
+  int max_iter = 100;
+  double tol = 1.0e-10;
+  int verbosity = Belos::Errors + Belos::Warnings;
+  if (print)
+    verbosity += Belos::TimingDetails + Belos::StatusTestDetails;
+
+  pList = Teuchos::rcp( new Teuchos::ParameterList );
+  pList->set( "Maximum Iterations", max_iter );
+  pList->set( "Convergence Tolerance", tol );
+  pList->set( "Verbosity", verbosity );
+
+  pBelos = Teuchos::rcp( new Belos::MinresSolMgr<double,MV,OP>(pProb, pList) );
+}
+
+template<class OP, class ST, class MP, class MV>
+void Iterative_Inverse_Operator<OP, ST, MP, MV>::operator () (const MV &b, MV &x)
+{
+  int pid = pComm->getRank();
+
+  // Initialize the solution to zero
+  x.putScalar( 0.0 );
+
+  // Reset the solver, problem, and status test for next solve (HKT)
+  pProb->setProblem( Teuchos::rcp(&x, false), Teuchos::rcp(&b, false) );
+
+  timer.start();
+  Belos::ReturnType ret = pBelos->solve();
+  timer.stop();
+
+  if (pid == 0 && print) {
+    if (ret == Belos::Converged)
+    {
+      std::cout << std::endl << "pid[" << pid << "] Minres converged" << std::endl;
+      std::cout << "Solution time: " << timer.totalElapsedTime() << std::endl;
+
+    }
+    else
+      std::cout << std::endl << "pid[" << pid << "] Minres did not converge" << std::endl;
+  }
+}
+
+//************************************************************************************************
+
+template <class ScalarType>
+int run(int argc, char *argv[]){
+  // Get default Tpetra template types
+  using ST = typename Tpetra::MultiVector<ScalarType>::scalar_type;
+  using LO = typename Tpetra::Vector<>::local_ordinal_type;
+  using GO = typename Tpetra::Vector<>::global_ordinal_type;
+  using NT = typename Tpetra::Vector<>::node_type;
+
+  // Init Tpetra types
+  using OP = typename Tpetra::Operator<ST,LO,GO,NT>;
+  using MV = typename Tpetra::MultiVector<ST,LO,GO,NT>;
+  using MP = typename Tpetra::Map<LO,GO,NT>;
+
+  using Teuchos::RCP;
+  using Teuchos::rcp;
+
+  Teuchos::GlobalMPISession session(&argc, &argv, NULL);
+
+  RCP<const Teuchos::Comm<int>> comm = Tpetra::getDefaultComm();
+  const int pid = rank(*comm);
+
+  bool verbose = false;
+  bool success = true;
+
+  try {
+
+    int n(10);
+    int numRHS=1;
+    
+    RCP<MP> map = RCP(new MP(n, 0, comm));
+    
+    MV X(map, numRHS), Y(map, numRHS);
+    X.putScalar( 1.0 );
+    
+    // Inner computes inv(D2)*y
+    RCP<Diagonal_Operator_2<MV>> D2 = rcp(new Diagonal_Operator_2<MV>(n, map->getMinGlobalIndex(), 1.0));
+    Iterative_Inverse_Operator<OP, ST, MP, MV> A2(n, 1, D2, "Belos (inv(D2))", true);
+    
+    // should return x=(1, 1/2, 1/3, ..., 1/10)
+    // AM: If you uncomment this line, test failed, I don't know why
+    // A2(X,Y);
```

Comment:
Issue here

Reply:
thanks @antoinemeyer5 . We at NGA will have to resolve this ourselves. I'll try to find time to take a look.

Reply:
Resolved